bot/exts/evergreen/youtube.py

```diff
@@ -0,0 +1,98 @@
+import logging
+from dataclasses import dataclass
+from html import unescape
+from typing import List
+from urllib.parse import quote_plus
+
+from discord import Embed
+from discord.ext import commands
+from discord.utils import escape_markdown
+
+from bot.constants import Colours, Tokens
+
+log = logging.getLogger(__name__)
+
+KEY = Tokens.youtube
+SEARCH_API = "https://www.googleapis.com/youtube/v3/search"
+YOUTUBE_VIDEO_URL = "https://www.youtube.com/watch?v={id}"
+YOUTUBE_SEARCH_URL = "https://www.youtube.com/results?search_query={search}"
+RESULT = "`{index}` [{title}]({url}) - {author}"
+
+
+@dataclass
+class SearchResult:
+    """Represents one video."""
+
+    title: str
+    author: str
+    id: str
+
+
+class YouTubeSearch(commands.Cog):
+    """Sends the top 5 results of a query from YouTube."""
+
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+        self.http_session = bot.http_session
+
+    async def search_youtube(self, search: str) -> List[SearchResult]:
+        """Queries API for top 5 results matching the search term."""
+        results = []
+        async with self.http_session.get(
+            SEARCH_API,
+            params={"part": "snippet", "q": search, "type": "video", "key": KEY},
+        ) as response:
+            if response.status != 200:
+                log.error("youtube response not succesful")
+                return None
+
+            data = await response.json()
+            for item in data["items"]:
+                results.append(
+                    SearchResult(
+                        title=escape_markdown(unescape(item["snippet"]["title"])),
```

Comment:
Well I don’t know about this, but what if the `link` (example) doesn’t exist it will catch a error then, we could make use of `try`(element doesn’t exist) or `or` (If element exists but is empty).

Reply:
@Shivansh-007 
If `snippet` or `title` or something else doesn't exist, it will throw an exception that will be caught at top-level.
That situation means that:
a) Google has changed their API, so this needs to be changed anyway
b) Google has a bug, which is pretty unlikely
(if that's what you're talking about)

Reply:
in other words, I don't see the value in checking for that.

Reply:
It does seem pretty unlikely, I don’t think so we need this, if we come across this while it is deployed the. We could add it.

Reply:
@Shivansh-007 
Exception handling in "user code" (implementation of commands) is needed when there can reasonably be an exception, which you can interpret as an error in the command input, or interpret as something useful. For example, if we had a `regex` command, we could interpret a failure to parse a string as a regular expression and report it to the user.

If, however:

- Google has changed their API
- There's a bug at Google
- There's a bug in Linux
- There's a bug in discord.py
- There's a bug in CPython
- You get `MemoryError` because you're out of memory
- Internet connection is broken
- The server is being melted by alien lasers

you can't really do much about it, so you leave it to the top-level handler.

Reply:
Ah right.